pkg/diagnostics/log/log.go

```diff
@@ -0,0 +1,280 @@
+package log
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	ct "github.com/daviddengcn/go-colortext"
+	"io"
+	"io/ioutil"
+	"runtime"
+	"strings"
+	"text/template"
+
+	"github.com/openshift/origin/pkg/version"
+)
+
+type LoggerOptions struct {
+	Level  int
+	Format string
+	Out    io.Writer
+}
+
+func (o *LoggerOptions) NewLogger() (*Logger, error) {
+	out := o.Out
+	if out == nil {
+		out = ioutil.Discard
+	}
+
+	ret, err := NewLogger(o.Level, o.Format, out)
+	return ret, err
+}
+
+type Level struct {
+	Level  int
+	Name   string
+	Prefix string
+	Color  ct.Color
+	Bright bool
+}
+
+func (l Level) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + l.Name + `"`), nil
+}
+
+func (l Level) MarshalYAML() (interface{}, error) {
+	return l.Name, nil
+}
+
+type Logger struct {
+	loggerType
+	level        Level
+	warningsSeen int
+	errorsSeen   int
+}
+
+// Internal type to deal with different log formats
+type loggerType interface {
+	Write(Entry)
+	Finish()
+}
+
+func NewLogger(setLevel int, setFormat string, out io.Writer) (*Logger, error) {
```

Comment:
Based on https://github.com/openshift/origin/pull/2049#issuecomment-131992506, we can simplify this.  We won't need json and yaml serialization to start.  That will let you eliminate the golang template calls.  Most people find them very hard to read when compared to printf.

I think that's fine because of the `DiagnosticResult` type that could later be used to include pertinent metadata for structured results if we ever develop that far.